go/vt/mysqlctl/capabilities.go

```diff
@@ -0,0 +1,248 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+ Detect server flavors and capabilities
+*/
+
+package mysqlctl
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	vtenv "vitess.io/vitess/go/vt/env"
+	"vitess.io/vitess/go/vt/log"
+)
+
+type serverVersion []int
+
+const (
+	flavorMySQL   = "mysql"
+	flavorPercona = "percona"
+	flavorMariaDB = "mariadb"
+)
+
+const (
+	CapabilityMySQLUpgradeInServer = 1 << iota
+	CapabilityInitializeInServer
+	CapabilityMySQLxEnabledByDefault
+	CapabilityPersistConfig
+	CapabilityShutdownCommand
+	CapabilityBackupLocks
+	CapabilityDefaultUtf8mb4
+	CapabilitySemiSyncEnabledByDefault
+	CapabilitySystemd // often package specific, not just version specific
+)
+
+var (
+	minimumMySQLUpgradeInServer            = serverVersion{8, 0, 16}
+	minimumMySQLInitializeInServer         = serverVersion{5, 7, 0}
+	minimumMySQLxEnabledByDefault          = serverVersion{8, 0, 11}
+	minimumMySQLPersistConfig              = serverVersion{8, 0, 0}
+	minimumMySQLShutdownCommand            = serverVersion{5, 7, 9}
+	minimumMySQLBackupLocks                = serverVersion{8, 0, 0}
+	minimumMySQLDefaultUtf8mb4             = serverVersion{8, 0, 0}
+	minimumMariaDBSemiSyncEnabledByDefault = serverVersion{10, 3, 3}
+	minimumMariaDBShutdownCommand          = serverVersion{10, 0, 4}
+)
+
+// HasCapability will return a bool value of if the server
+// has a particular feature, based on a known list of
+// introduced versions and detection. This gets more
+// complicated in MySQL 8.0 as several features were introduced
+// after the initial GA release.
+func (mysqld *Mysqld) HasCapability(capability int) bool {
+
+	switch capability {
+	case CapabilityMySQLUpgradeInServer:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLUpgradeInServer)
+	case CapabilityInitializeInServer:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLInitializeInServer)
+	case CapabilityMySQLxEnabledByDefault:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLxEnabledByDefault)
+	case CapabilityPersistConfig:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLPersistConfig)
+	case CapabilityShutdownCommand:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLShutdownCommand) || mysqld.IsMariaDB() && mysqld.greaterThan(minimumMariaDBShutdownCommand)
+	case CapabilityBackupLocks:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLBackupLocks)
+	case CapabilityDefaultUtf8mb4:
+		return mysqld.IsMySQLLike() && mysqld.greaterThan(minimumMySQLDefaultUtf8mb4)
+	case CapabilitySemiSyncEnabledByDefault:
+		return mysqld.IsMariaDB() && mysqld.greaterThan(minimumMariaDBSemiSyncEnabledByDefault)
+	case CapabilitySystemd:
+		dir, err := vtenv.VtMysqlRoot()
+		if err != nil {
+			return false
+		}
+		_, err = binaryPath(dir, "mysqld_safe")
+		return err != nil // if there is no mysqld_safe, it means this platform is systemd
```

Comment:
I've removed mysqld_safe from the capabilities system, and reverted this part of the code. I think it will need a longer-term solution but whether mysqld_safe is available is orthogonal to the server version.